### PR TITLE
Add parent and child details in snapshot list command

### PR DESCRIPTION
**Pull request by slalwani97**
**What this PR does / why we need it**:
* Adds template to display snapshots.
* Adds Parent and Children details in snapshot list command
* Adds unit test cases in snapshot_test.go file

**Which issue this PR fixes**: <https://github.com/openebs/openebs/issues/1604>

**Special notes for your reviewer**:
screenshot from 2018-06-12 12-32-52

### CODE REVIEW

File: pkg/client/mapiserver/snapshot_test.go [hunk 1]
```diff
@@ -14,11 +14,13 @@ import (
...
+	snapshotResponse      = `{"actions":{},"id":"snapdemo1","links":{"self":"http://10.36.0.1:9501/v1/snapshotoutputs/snapdemo1"},"type":"snapshotOutput"}`
+	volumeNameIsMissing   = errors.New("Volume name is missing")
+	badReqErr             = errors.New(snapshotResponse)
+	volNotFound           = errors.New("Volume not found")
+	SnapshotListResponse  = `{"volume-snap-snap1.img": {"name": "volume-snap-snap1.img", "parent": "", "children":[ "volume-snap-snap2.img", "volume-head-001.img"], "usercreated":true, "removed":false, "created": "2018-06-10T19:33:34Z", "size": "0"}, "volume-snap-snap2.img": {"name": "volume-snap-snap2.img", "parent": "volume-snap-snap1.img", "children":[ "volume-snap-snap3.img"], "created": "2018-06-10T19:33:34Z", "size": "0"}, "volume-snap-snap3.img": {"name": "volume-snap-snap3.img", "parent": "", "children":[], "usercreated":true, "removed":false, "created": "2018-06-10T19:33:34Z", "size": "0"}}`
+	SnapshotListResponse1 = `{"volume-head-001.img": {"name": "volume-head-001.img", "parent": "", "children":[], "usercreated":true, "removed":false, "created": "2018-06-10T19:33:34Z", "size": "0"}}`
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/snapshot_test.go]: SnapshotListResponse1 --> NoSnapshot ?

Comment (slalwani97) [reply pkg/client/mapiserver/snapshot_test.go]: done..

File: pkg/client/mapiserver/snapshot.go [hunk 2]
```diff
@@ -34,9 +35,17 @@ const (
 	http_timeout = 5 * time.Second
 )
 
+// SnapshotInfo stores the details of snapshot
+type SnapshotInfo struct {
+	Name     string
+	Created  string
+	Size     string
+	Parent   string
```
Comment (kmova) [inline pkg/client/mapiserver/snapshot.go]: Can you add comments on what is snapshot parent and child?

**Comment by slalwani97** [comment]
screenshot from 2018-06-13 13-15-04

@kmova Sir, multiple children will be displayed in multiple rows.

**Comment by kmova** [comment]
Thanks @slalwani97. Looking at the output.. it might help if we have a tree kind of display for this. 

```
snap-1
  snap-2
     snap-3
     head-010
```

**Comment by slalwani97** [comment]
```
Snapshot Details: 
------------------
NAME     CREATED AT                      SIZE     PARENT     CHILDREN
-----    -----------                     -----    -------    --------- 
snap1    Wed Jun 13 19:46:45 UTC 2018    0        NA         head-009
                                                             snap2
                                                              
snap2    Wed Jun 13 19:47:29 UTC 2018    0        snap1      snap3
                                                              
snap3    Wed Jun 13 19:47:31 UTC 2018    0        snap2      snap4
                                                              
snap4    Wed Jun 13 19:47:34 UTC 2018    0        snap3      snap5
                                                              
snap5    Wed Jun 13 19:47:36 UTC 2018    0        snap4      NA
```

@kmova How about this? 
The order of snapshots will be according to the time created.

**Comment by kmova** [comment]
@slalwani97 - The above output looks good. 

One more question though. head-009 -- why is this showing as a child to snap1 - did you do a snapshot restore to snap1 ?

**Comment by slalwani97** [comment]
@kmova I have reverted the current state to snap1.

**Comment by slalwani97** [comment]
```
Snapshot Details: 
------------------
NAME              CREATED AT                      SIZE         PARENT            CHILDREN
-----             -----------                     -----        -------           --------- 
pvc-snap1-test    Thu Jun 14 09:29:49 UTC 2018    393015296    NA                pvc-snap2-test
                                                                                  
pvc-snap2-test    Thu Jun 14 09:30:24 UTC 2018    0            pvc-snap1-test    pvc-snap3-test
                                                                                  
pvc-snap3-test    Thu Jun 14 09:30:41 UTC 2018    0            pvc-snap2-test    pvc-snap4-test
                                                                                  
pvc-snap4-test    Thu Jun 14 09:30:44 UTC 2018    0            pvc-snap3-test    head-004
                                                                                  
```

**Comment by utkarshmani1997** [comment]
@slalwani97 you are doing great, i just looked at @kmova 's comment and while browsing i discovered [this](https://www.youtube.com/watch?v=XbKSssBftLM) today. Hope this video help you in implementing the output in tree format.

File: pkg/client/mapiserver/snapshot.go [hunk 3]
```diff
@@ -196,3 +239,42 @@ func getInfo(body []byte) (map[string]client.DiskInfo, error) {
...
+	err := tmpl.Execute(w, snapshotList)
+	if err != nil {
+		return err
+	}
+	w.Flush()
+
+	return nil
+}
+
+func sortSnapshotDisksByDateTime(snapshotDisks map[int]*SnapshotInfo) {
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/snapshot.go]: Better to put these helper functions in util,go file. ?

File: pkg/client/mapiserver/snapshot.go [hunk 4]
```diff
@@ -196,3 +239,42 @@ func getInfo(body []byte) (map[string]client.DiskInfo, error) {
...
+	sort.Sort(SortByDateTime(snapshotDisks))
+}
+
+func changeDateFormatToUnixDate(snapshotDisks map[int]*SnapshotInfo) error {
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/snapshot.go]: Same as above.

File: pkg/client/mapiserver/utils.go [hunk 5]
```diff
@@ -39,3 +41,24 @@ func getDefaultAddr() string {
 	}
 	return "http://" + env + ":5656"
 }
+
+// SortSnapshotDisksByDateTime orders the snapshot disks with respect to date and time
+func SortSnapshotDisksByDateTime(snapshotDisks []SnapshotInfo) {
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/utils.go]: Can we write unit test for this ?

File: pkg/client/mapiserver/snapshot.go [hunk 6]
```diff
@@ -196,3 +230,27 @@ func getInfo(body []byte) (map[string]client.DiskInfo, error) {
 	}
 	return s, err
 }
+
+func displayVolumeSnapshot(snapshotList []SnapshotInfo) error {
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/snapshot.go]: Can we write unit test for this ?

File: pkg/client/mapiserver/utils.go (hunk 5 above)
Comment (slalwani97) [reply pkg/client/mapiserver/utils.go]: This function is already covered while testing the main function i.e ListSnapshot.

File: pkg/client/mapiserver/snapshot.go (hunk 6 above)
Comment (prateekpandey14) [reply pkg/client/mapiserver/snapshot.go]: cool..

**Pull request merged**